File: bench/fnv1a64.py
# BOT-ASSISTED

import sys


CAP = 65536
RUNS = 2048
OFFSET = 0xCBF29CE484222325
PRIME = 0x100000001B3
MASK = 0xFFFFFFFFFFFFFFFF
WANT = 0xB0A1EA8560222325
OFFSET_HI = 0xCBF29CE4
OFFSET_LO = 0x84222325
PRIME_LO = 0x1B3
BUF = (b"0123456789abcdef" * (CAP // 16))


def fnv1a64_int(hash: int, src: bytes) -> int:
    prime = PRIME
    mask = MASK
    for byte in src:
        hash ^= byte
        hash = (hash * prime) & mask
    return hash


def fnv1a64_u32(hi: int, lo: int, src: bytes) -> tuple[int, int]:
    prime = PRIME_LO
    for byte in src:
        lo = (lo ^ byte) & 0xFFFFFFFF
        carry = (((lo >> 16) * prime) + (((lo & 0xFFFF) * prime) >> 16)) >> 16
        hi = ((hi * prime) + carry + ((lo << 8) & 0xFFFFFFFF)) & 0xFFFFFFFF
        lo = (lo * prime) & 0xFFFFFFFF
    return hi, lo


def main() -> None:
    if sys.implementation.name == "pypy":
        hi = OFFSET_HI
        lo = OFFSET_LO
        for _ in range(RUNS):
            hi, lo = fnv1a64_u32(hi, lo, BUF)
        hash = (hi << 32) | lo
    else:
        hash = OFFSET
        for _ in range(RUNS):
            hash = fnv1a64_int(hash, BUF)
    if hash != WANT:
        raise RuntimeError(f"mismatch: expected {WANT:#x}; got {hash:#x}")


if __name__ == "__main__":
    main()

File: bench/fnv1a64.java
// BOT-TRANSLATED

final class fnv1a64 {
  static final long OFFSET = 0xcbf29ce484222325L;
  static final long PRIME = 0x100000001b3L;
  static final long WANT = 0xb0a1ea8560222325L;
  static volatile int capacity = 65536;
  static volatile int runs = 2048;

  static void init(byte[] buf) {
    final String pattern = "0123456789abcdef";

    for (int ind = 0; ind < buf.length; ind++) {
      buf[ind] = (byte) pattern.charAt(ind & 15);
    }
  }

  static long hash(long hash, byte[] src) {
    for (final byte val : src) {
      hash ^= val & 0xffL;
      hash *= PRIME;
    }

    return hash;
  }

  static void run(int capacity, int runs) {
    final byte[] buf = new byte[capacity];
    init(buf);

    long hash = OFFSET;
    while (runs-- > 0) hash = hash(hash, buf);

    if (hash != WANT) {
      throw new AssertionError(
        "mismatch: expected 0x"
          + Long.toHexString(WANT)
          + "; got 0x"
          + Long.toHexString(hash)
      );
    }
  }

  public static void main(String[] args) {
    run(capacity, runs);
  }
}
